### PR TITLE
Docker in Workspaces

### PR DESCRIPTION
**Pull request by csweichel**
This PR introduces `docker-up`, a utility to start `dockerd` in user-namespaced workspaces.

TLDR; together with user-named workspaces and this PR you can now have Docker in a workspace with:
```
sudo docker-up
```

### What it does
- checks if `dockerd` and `slirp4netns` are installed. If not, and `apt-get` is available, it installs the `docker.io` and `slirp4netns` packages.
- clone's itself in a new network namespace to give docker "free reign" (aka "the child process")
- starts `slirp4netns` to connect the network namespace to the outside world, and tells the child process that things are set up
- the child process starts `dockerd` with a custom container runtime (`runc-facade`) and userland proxy (`slirp-docker-proxy`)
- once the daemon is running, the child process runs `chmod 666 /var/run/docker.socket` to make the docker daemon available to everyone in the workspace (enables `docker run` as compared to `sudo docker run`).

We adds those three binaries to the supervisor layer, which makes them available in `/theia`, or `/usr/bin` when registry-facade is enabled.

`runc-facade` deals with some idiosyncrasies of running rootless. It modifies the OCI runtime spec of the new container to
  - make `/proc` and `/sys` work (currently replacing it with a bind mount to the workspace's `/proc`/`/sys` - far from ideal, but prepwork for when we have proper `/proc` mounts).
  - set `oomScoreAdj` to 1000, which is above the oom score of the workspace container, and hence doesn't require `CAP_SYS_RESOURCE` in the root namespace ([fs/proc/base.c](https://github.com/torvalds/linux/blob/c4d6fe7311762f2e03b3c27ad38df7c40c80cc93/fs/proc/base.c#L1081-L1082))
  - register [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#posix-platform-hooks) to actually establish/clean up the proc/sys mounts in the future

`slirp-docker-proxy` works much the same way as [moby's rootlesskit support does it](https://github.com/moby/moby/pull/38913). docker-up registers it as docker userland proxy which is started for each exposed port. We then talk to slirp4net directly to add/remove the port. (With rootlesskit, the userland proxy would talk to rootlesskit, which in turn would talk to slirp4net.)

This PR also makes several extensions to the custom seccomp profile to facilitate running Docker.
Those changes by themselves should not be exploitable, but do remove some "defense-in-depth" layer. We might want to build that back up in the future.

### How to test
- login to https://cw-docker-up.staging.gitpod-dev.com/workspaces
- give your user the `privileged-ws` permissions and `user_namespace` feature flag by running
  ```
  leeway run components/ws-daemon:enabled-userns-ff
  ```
- start a workspace
- run `sudo /theia/docker-up`
- in a new terminal, interact with docker or docker-compose, e.g. `docker run --rm -it alpine:latest`

Building gitpod-io docker images works with this method:
image

### CODE REVIEW

**Comment by akosyakov** [comment]
I tried to test first without `user_namespace` feature flag and was not able to start any workspace:
Screenshot 2020-11-03 at 14 37 37

**Comment by csweichel** [comment]
On one node the ws-daemon did not come up because the port was already taken.
That happens occasionally on dev-staging, as we're just rolling dice for those ports.

I've restarted the deployment, and now the daemon starts on all nodes.

**Comment by akosyakov** [comment]
Nice I was able to preview the app started from the docker compose ❤️ : 
Screenshot 2020-11-03 at 16 30 08

**Review by akosyakov — APPROVED** [review]
It worked nicely besides me forgetting to use `sudo` here and there :D

**Review by JesterOrNot — COMMENTED** [review]
Docker exec does not work

Errors with

```
gitpod /workspace/Microservice-Thing $ sudo docker exec optimistic_herschel ls
sudo: setrlimit(RLIMIT_CORE): Operation not permitted
OCI runtime exec failed: exec failed: container_linux.go:349: starting container process caused "process_linux.go:101: executing setns process caused \"exit status 1\"": unknown
```

Reproduction steps

1. start a docker container e.g. `docker run -it alpine`
2. open new terminal
3. Get the name of the created container via `docker container ls`
4. try `docker exec <container name> ls`

**Comment by csweichel** [comment]
added an issue for `docker exec`: https://github.com/gitpod-io/gitpod/issues/2123
